### PR TITLE
add select_any impl

### PR DESCRIPTION
**Pull request by bluejekyll**
@alexcrichton this is the select_any as discussed.

I got rid of the index from the result, b/c it didn't seem worth it to track for the last error. It was an issue due to the skipping of errors, which result in those being removed from the inner Vec.

I added the loop { ... } so this should be good. Otherwise, still pretty close to the select_all interface.

### CODE REVIEW

File: src/select_any.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    inner: iter.into_iter()
+    .map(|a| a.into_future())
+    .collect(),
+  };
+  assert!(ret.inner.len() > 0);
+  ret
+}
+
+impl<A> Future for SelectAny<A> where A: Future {
+  type Item = (A::Item, Vec<A>);
```
Comment (alexcrichton) [inline src/select_any.rs]: Mind bumping these up to 4-space tabs?

Comment (alexcrichton) [inline src/select_any.rs]: Ah I personally avoid this in repos for now, but when rustfmt comes along hopefully that'll help!

**Comment by alexcrichton** [comment]
Ok, my only remaining qualm here is the name. I feel like `select_any` is a little too similar to `select_all` to move forward with this naming unfortunately. That is, it's kind of unclear why `any` might wait for all futures (currently) where `all` will only ever wait for at most one future.

Maybe something like `select_ok`, though, to emphasize that it explicitly tries to select an `Ok` variant?

**Comment by bluejekyll** [comment]
Yeah, i've been thinking about the naming for a little bit. The reasoning behind the `any` was derived from `any` in Iterator. It's basically meant to say, select any ok result, otherwise it is err. This is the same definition of `any` in Iterator, that is it is true if anything in the iterator is true.

Now the complicated thing is that in this context `select_all`, means select across all futures in a vector and return the first that completes, which matches the definition of `select`. `all`, though, does not necessarily fit this context, at least not as defined on Iterator, which requires that all elements must be successful. For that in futures-rs, there is `collect`, but `collect` does not function like `select`, and also is implemented as requiring each `Future` to complete in sequence, not in parallel like `select`. This leads me to think that `select_all` is possibly misnamed as well :). It also makes me think that `join_all` is possibly missing as well, though the location I would possibly use it is actually better implemented with a `select_all` with a loop since I need to process all results, doing something with errors and successes...

Anyway, long story short, I'm happy to change the name, perhaps dropping `select` is the correct option, because given current semantics, that's the one that `select_any` matches least. What do you think of just a bare `any`? or `any_ok`

**Comment by carllerche** [comment]
`select_ok` maybe?

**Comment by alexcrichton** [comment]
How about `first_ok`? I tried to keep the name "select" with `select_all` to be similar to `Future::select`, but it's true that `collect` should actually be `join_all` if that's the case. I don't mind keeping the "breakaway" though from the standard names on `Future` to get a more invocative name.

**Comment by bluejekyll** [comment]
I'm leaning towards `first_ok` but `select_ok` is good too.

It's your library, you decide ;)

**Comment by alexcrichton** [comment]
Ok, let's go with `select_ok` as I'd probably like to rename `collect` to `join_all` to keep a similar pattern there.

**Comment by alexcrichton** [comment]
Thanks!
